File: spec/file_buffer_spec.rb
# -*- encoding: utf-8 -*-
require File.expand_path("../spec_helper", __FILE__)
require 'fileutils'

describe Backchat::Hookup::FileBuffer do
  
  context "when opening" do
    work_path = "./rb-test-work"
    log_path = "./rb-test-work/testing/and/such/buffer.log"

    before(:each) do 
      FileUtils.rm_rf(work_path) if File.exist?(work_path)
    end

    it "creates the path if missing" do
      buff = FileBuffer.new(work_path)
      buff.on(:open) do
        File.exist?(work_path).should be_true
      end
      buff.open
      FileUtils.rm_rf(work_path)
    end
  end

  context 'when not draining' do
    work_path = "./rb-test-work2"
    log_path = "#{work_path}/buffer.log"
    exp1 = "the first message"
    exp2 = "the second message"

    before(:each) do
      FileUtils.rm_rf(work_path) if File.exist?(work_path)
    end

    after(:each) do 
      FileUtils.rm_rf(work_path)
    end

    it "writes to a file" do      
      buff = FileBuffer.new(log_path)
      lines = []
      buff.open
      buff.write(exp1)
      buff.write(exp2)
      buff.close
      File.open(log_path).readlines.size.should == 2      
    end
  end 

  context "when draining" do
    work_path = "./rb-test-work3"
    log_path = "#{work_path}/buffer.log"
    exp1 = "the first message"
    exp2 = "the second message"

    before(:each) do
      FileUtils.rm_rf(work_path) if File.exist?(work_path)
    end

    after(:each) do 
      FileUtils.rm_rf(work_path)
    end

    it "writes new sends to the memory buffer" do
      mem = []
      buff = FileBuffer.new(log_path, :memory_buffer => mem, :initial_state => 1)
      buff.write(exp1)
      buff.write(exp2)
      buff.close
      mem.size.should == 2     
    end

    it "raises data events for every line in the buffer" do
      lines = []
      buff = FileBuffer.new(log_path)
      buff.on(:data) do |args|
        lines << args
      end
      buff.open
      buff.write(exp1)
      buff.write(exp2) do
        buff.drain
      end
      buff.close
      lines.size.should == 2
    end
  end
end

File: src/test/scala/io/backchat/hookup/tests/FileBufferSpec.scala
package io.backchat.hookup
package tests

import org.specs2.specification.AfterAll
import org.specs2.time.NoTimeConversions
import org.specs2.Specification
import java.io.File
import org.apache.commons.io.{FilenameUtils, FileUtils}
import org.json4s._
import scala.io.Source
import collection.JavaConverters._
import org.specs2.specification.core.{Fragments}
import java.util.concurrent.{Executors, ConcurrentLinkedQueue}
import scala.concurrent.{Await, Future, ExecutionContext}
import scala.concurrent.duration._
import akka.actor.ActorSystem
import collection.mutable.{ArrayBuffer, Buffer, SynchronizedBuffer, ListBuffer}
import java.util.concurrent.atomic.AtomicInteger

class FileBufferSpec extends Specification with AfterAll { def is =
  "A FileBuffer should" ^
    "create the path to the file if it doesn't exist" ! createsPath ^
    "write to a file while the buffer is open" ! writesToFile ^
    "write to memory buffer while draining" ! writesToMemory ^
    "drain the buffers" ! drainsBuffers ^
    "not fail under concurrent load" ! handlesConcurrentLoads ^
  end

  implicit val wireFormat: WireFormat = new JsonProtocolWireFormat()(DefaultFormats)
  implicit val executionContext = ExecutionContext.fromExecutorService(Executors.newCachedThreadPool())


  override def afterAll(): Unit = executionContext.shutdown()

  def createsPath = {
    val logPath = new File("./test-work/testing/and/such/buffer.log")
    val workPath = new File("./test-work")
    if (workPath.exists()) FileUtils.deleteDirectory(workPath)
    val buff = new FileBuffer(logPath)
    buff.open()
    val res = logPath.getParentFile.exists must beTrue
    FileUtils.deleteDirectory(workPath)
    buff.close()
    res
  }

  def writesToFile = {
    FileUtils.deleteQuietly(new File("./test-work2"))
    val logPath = new File("./test-work2/buffer.log")
    val buff = new FileBuffer(logPath)
    val exp1: OutboundMessage = TextMessage("the first message")
    val exp2: OutboundMessage = TextMessage("the second message")
    buff.open()
    buff.write(exp1)
    buff.write(exp2)
    buff.close()
    val lines = Source.fromFile(logPath).getLines().toList map wireFormat.parseOutMessage
    FileUtils.deleteQuietly(new File("./test-work2"))
    lines must contain(eachOf(exp1, exp2))
  }

  def writesToMemory = {
    val logPath = new File("./test-work3/buffer.log")
    val exp1: OutboundMessage = TextMessage("the first message")
    val exp2: OutboundMessage = TextMessage("the second message")
    val queue = new ConcurrentLinkedQueue[String]()
    val buff = new FileBuffer(logPath, false, queue)
    buff.open()
    buff.write(exp1)
    buff.write(exp2)
    val lst = queue.asScala.toList
    buff.close()
    FileUtils.deleteDirectory(new File("./test-work3"))
    lst must contain(eachOf(wireFormat.render(exp1), wireFormat.render(exp2)))
  }

  def drainsBuffers = {
    val logPath = new File("./test-work4/buffer.log")
    val buff = new FileBuffer(logPath)
    val exp1: OutboundMessage = TextMessage("the first message")
    val exp2: OutboundMessage = TextMessage("the second message")
    buff.open()
    buff.write(exp1)
    buff.write(exp2)
    val lines = new ListBuffer[OutboundMessage]
    Await.ready(buff drain { out =>
      Future {
        lines += out
        Success
      }
    }, 5 seconds)
    buff.close()
    FileUtils.deleteQuietly(new File("./test-work4"))
    lines must contain(eachOf(exp1, exp2))
  }

  def handlesConcurrentLoads = {
    val system = ActorSystem("filebufferconc")
    val logPath = new File("./test-work5/buffer.log")
    val buff = new FileBuffer(logPath)
    val lines = new ArrayBuffer[OutboundMessage] with SynchronizedBuffer[OutboundMessage]
    buff.open()
    val reader = system.scheduler.schedule(50 millis, 50 millis) {
      Await.ready(buff drain { out =>
        Future {
          lines += out
          Success
        }
      }, 5 seconds)
    }
    (1 to 20000) foreach { s =>
      buff.write(TextMessage("message %s" format s))
    }
    reader.cancel()
    Await.ready(buff drain { out =>
          Future {
            lines += out
            Success
          }
        }, 5 seconds)
    buff.close()
    FileUtils.deleteDirectory(new File("./test-work5"))
    system.shutdown()
    lines must haveSize(20000)
  }

}
